p2p/src/protocol.rs

```diff
@@ -50,223 +45,188 @@ impl Protocol {
 				if let Some(b) = bo {
-					return Ok(Some(Msg::new(Type::Block, b, self.peer_info.version)?));
+					Consumed::Response(Msg::new(Type::Block, b, self.peer_info.version)?)
+				} else {
+					Consumed::None
 				}
-				Ok(None)
 			}
 
-			Type::Block => {
-				debug!(
-					"handle_payload: received block: msg_len: {}",
-					msg.header.msg_len
-				);
-				let b: core::UntrustedBlock = msg.body()?;
-
+			Message::Block(b) => {
+				debug!("handle_payload: received block");
 				// We default to NONE opts here as we do not know know yet why this block was
 				// received.
 				// If we requested this block from a peer due to our node syncing then
 				// the peer adapter will override opts to reflect this.
 				adapter.block_received(b.into(), &self.peer_info, chain::Options::NONE)?;
-				Ok(None)
+				Consumed::None
 			}
 
-			Type::GetCompactBlock => {
-				let h: Hash = msg.body()?;
+			Message::GetCompactBlock(h) => {
 				if let Some(b) = adapter.get_block(h, &self.peer_info) {
 					let cb: CompactBlock = b.into();
-					Ok(Some(Msg::new(
-						Type::CompactBlock,
-						cb,
-						self.peer_info.version,
-					)?))
+					Consumed::Response(Msg::new(Type::CompactBlock, cb, self.peer_info.version)?)
 				} else {
-					Ok(None)
+					Consumed::None
 				}
 			}
 
-			Type::CompactBlock => {
-				debug!(
-					"handle_payload: received compact block: msg_len: {}",
-					msg.header.msg_len
-				);
-				let b: core::UntrustedCompactBlock = msg.body()?;
-
+			Message::CompactBlock(b) => {
+				debug!("handle_payload: received compact block");
 				adapter.compact_block_received(b.into(), &self.peer_info)?;
-				Ok(None)
+				Consumed::None
 			}
 
-			Type::GetHeaders => {
+			Message::GetHeaders(loc) => {
 				// load headers from the locator
-				let loc: Locator = msg.body()?;
 				let headers = adapter.locate_headers(&loc.hashes)?;
 
 				// serialize and send all the headers over
-				Ok(Some(Msg::new(
+				Consumed::Response(Msg::new(
 					Type::Headers,
 					Headers { headers },
 					self.peer_info.version,
-				)?))
+				)?)
 			}
 
 			// "header first" block propagation - if we have not yet seen this block
 			// we can go request it from some of our peers
-			Type::Header => {
-				let header: core::UntrustedBlockHeader = msg.body()?;
+			Message::Header(header) => {
 				adapter.header_received(header.into(), &self.peer_info)?;
-				Ok(None)
+				Consumed::None
 			}
 
-			Type::Headers => {
-				let mut total_bytes_read = 0;
-
-				// Read the count (u16) so we now how many headers to read.
-				let (count, bytes_read): (u16, _) = msg.streaming_read()?;
-				total_bytes_read += bytes_read;
-
+			Message::Headers(hs) => {
 				// Read chunks of headers off the stream and pass them off to the adapter.
-				let chunk_size = 32u16;
-				let mut headers = Vec::with_capacity(chunk_size as usize);
-				for i in 1..=count {
-					let (header, bytes_read) =
-						msg.streaming_read::<core::UntrustedBlockHeader>()?;
```

Comment:
Related to my previous comment - here we used to actually read headers of the bytes from the body in chunks of 32 headers.
We now (I think?) read and parse _all_ headers up front and _then_ process them in chunks of 32.

Reply:
Related - looks like our 32 header chunks are acting weird here also (first chunk is a single header...) - 

```
20200914 18:07:40.223 INFO grin_servers::common::adapters - Received 1 block headers from 47.90.208.51:3414
20200914 18:07:40.251 INFO grin_servers::common::adapters - Received 32 block headers from 47.90.208.51:3414
20200914 18:07:40.304 INFO grin_servers::common::adapters - Received 32 block headers from 47.90.208.51:3414
20200914 18:07:40.355 INFO grin_servers::common::adapters - Received 32 block headers from 47.90.208.51:3414
20200914 18:07:40.410 INFO grin_servers::common::adapters - Received 32 block headers from 47.90.208.51:3414
20200914 18:07:40.461 INFO grin_servers::common::adapters - Received 32 block headers from 47.90.208.51:3414
20200914 18:07:40.513 INFO grin_servers::common::adapters - Received 32 block headers from 47.90.208.51:3414
20200914 18:07:40.568 INFO grin_servers::common::adapters - Received 32 block headers from 47.90.208.51:3414
20200914 18:07:40.620 INFO grin_servers::common::adapters - Received 32 block headers from 47.90.208.51:3414
20200914 18:07:40.674 INFO grin_servers::common::adapters - Received 32 block headers from 47.90.208.51:3414
20200914 18:07:40.726 INFO grin_servers::common::adapters - Received 32 block headers from 47.90.208.51:3414
20200914 18:07:40.778 INFO grin_servers::common::adapters - Received 32 block headers from 47.90.208.51:3414
20200914 18:07:40.830 INFO grin_servers::common::adapters - Received 32 block headers from 47.90.208.51:3414
20200914 18:07:40.881 INFO grin_servers::common::adapters - Received 32 block headers from 47.90.208.51:3414
20200914 18:07:40.928 INFO grin_servers::common::adapters - Received 32 block headers from 47.90.208.51:3414
20200914 18:07:40.980 INFO grin_servers::common::adapters - Received 32 block headers from 47.90.208.51:3414
20200914 18:07:41.036 INFO grin_servers::common::adapters - Received 31 block headers from 47.90.208.51:3414
```

Reply:
Maybe we don't actually care about the "streaming" approach. 
I'm not entirely sure how much benefit it gave us for a non-trivial amount of additional complexity.